core/src/main/java/io/questdb/cutlass/line/tcp/IOTableUpdateDetailsPool.java

```diff
@@ -0,0 +1,135 @@
+
+    public IOTableUpdateDetailsPool(int maxSize) {
+        pooledSize = maxSize;
+        createNewTableUpdateDetails = k -> new TableUpdateDetails[pooledSize];
+    }
+
+    @Override
+    public void close() throws IOException {
+        closed = true;
+        for (CharSequence key : pool.keySet()) {
+            final TableUpdateDetails[] tudArray = pool.get(key);
+            for (int n = tudArray.length, i = 0; i < n; i++) {
+                TableUpdateDetails tud = getVolatile(tudArray, i);
+                if (tud != null) {
+                    releaseTud(tud, null);
+                }
+            }
+        }
+    }
+
+    public void closeIdle(long nowMillis, long idleTimeout, LineTcpReceiver.SchedulerListener listener) {
+        for (CharSequence key : pool.keySet()) {
+            final TableUpdateDetails[] tudArray = pool.get(key);
+            if (tudArray != null) {
+                for (int n = tudArray.length, i = 0; i < n; i++) {
+                    TableUpdateDetails tud = getVolatile(tudArray, i);
+                    if (tud != null && nowMillis - tud.getLastMeasurementMillis() >= idleTimeout) {
+                        if (casTabAt(tudArray, i, tud, null)) {
+                            LOG.info().$("active table going idle [tableName=").$(tud.getTableNameUtf16()).I$();
+                            releaseTud(tud, listener);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public TableUpdateDetails get(CharSequence tableNameUtf8) {
+        TableUpdateDetails[] tudArray = pool.get(tableNameUtf8);
+        if (tudArray != null) {
+            for (int n = tudArray.length, i = 0; i < n; i++) {
+                TableUpdateDetails tud = getVolatile(tudArray, i);
+                if (tud != null && casTabAt(tudArray, i, tud, null)) {
+                    return tud;
+                }
+            }
+        }
+        return null;
+    }
+
+    public void put(CharSequence tableNameUtf8, TableUpdateDetails tud) {
+        TableUpdateDetails[] tudArray = pool.computeIfAbsent(tableNameUtf8, createNewTableUpdateDetails);
+        while (true) {
+            if (closed) {
+                releaseTud(tud, null);
+                return;
+            }
+            // There must be a place to return the tud, arrays are pre-allocate to hold enough for all IO threads
+            for (int n = tudArray.length, i = 0; i < n; i++) {
+                if (casTabAt(tudArray, i, null, tud)) {
+                    return;
+                }
+            }
+        }
+    }
+
+    private void releaseTud(TableUpdateDetails tud, LineTcpReceiver.SchedulerListener listener) {
+        tud.releaseWriter(true);
+        Misc.free(tud);
+        if (listener != null) {
+            listener.onEvent(tud.getTableToken(), 1);
+        }
+    }
+
+    static boolean casTabAt(TableUpdateDetails[] tab, int i, TableUpdateDetails expected, TableUpdateDetails v) {
```

Comment:
why not std AtomicReferenceArray? is it faster?

Reply:
It's same, can be `AtomicReferenceArray` but I already wrote it this way.